pkg/common/consumer/consumer_manager.go

```diff
@@ -0,0 +1,391 @@
+func (m *kafkaConsumerGroupManagerImpl) stopConsumerGroup(groupId string) error {
+	groupLogger := m.logger.With(zap.String("GroupId", groupId))
+	groupLogger.Info("Stopping Managed ConsumerGroup")
+	managedGrp, ok := m.groups[groupId]
+	if ok {
+		// The managedGroup's start channel must be created (that is, the managedGroup must be marked
+		// as "stopped") before closing the internal ConsumerGroup, otherwise the consume function would
+		// return control to the factory.
+		managedGrp.createRestartChannel()
+		// Close the inner sarama ConsumerGroup, which will cause our consume() function to stop
+		// and wait for the managedGroup to start again.
+		err := managedGrp.group.Close()
+		if err != nil {
+			groupLogger.Error("Failed To Close Managed ConsumerGroup", zap.Error(err))
+			// Don't leave the start channel open if the group.Close() call failed; that would be misleading
+			managedGrp.closeRestartChannel()
+			return err
+		}
+		return nil
+	}
+	groupLogger.Info("ConsumerGroup Not Managed - Ignoring Stop Request")
+	return fmt.Errorf("stop requested for consumer group not in managed list: %s", groupId)
+}
+
+// startConsumerGroups creates a new Consumer Group based on the groupId provided
+func (m *kafkaConsumerGroupManagerImpl) startConsumerGroup(groupId string) error {
+	groupLogger := m.logger.With(zap.String("GroupId", groupId))
+	groupLogger.Info("Starting Managed ConsumerGroup")
+	m.groupLock.Lock() // Don't allow m.groups to be modified while processing a start request
+	defer m.groupLock.Unlock()
+	if _, ok := m.groups[groupId]; ok {
+		group, err := m.factory.createConsumerGroup(groupId)
+		if err != nil {
+			groupLogger.Error("Failed To Restart Managed ConsumerGroup", zap.Error(err))
+			return err
+		}
+		m.groups[groupId].group = group
+		m.groups[groupId].closeRestartChannel() // Closing this allows the waitForStart function to finish
+		return nil
+	}
+	groupLogger.Info("ConsumerGroup Not Managed - Ignoring Start Request")
+	return fmt.Errorf("start requested for consumer group not in managed list: %s", groupId)
+}
+
+// processAsyncGroupNotification calls the provided groupFunction with whatever GroupId is contained
+// in the commandMessage, after verifying that the command version is correct.  It then calls the
+// appropriate Async response function on the commandMessage (NotifyFailed or NotifySuccess)
+func processAsyncGroupNotification(commandMessage ctrlservice.AsyncCommandMessage, groupFunction func(groupId string) error) {
+	if cmd, ok := commandMessage.ParsedCommand().(*commands.ConsumerGroupAsyncCommand); ok {
+		if cmd.Version != commands.ConsumerGroupAsyncCommandVersion {
+			commandMessage.NotifyFailed(fmt.Errorf("version mismatch; expected %d but got %d", commands.ConsumerGroupAsyncCommandVersion, cmd.Version))
+		} else {
+			// Calling NotifyFailed with a nil error is the same as calling NotifySuccess
+			commandMessage.NotifyFailed(groupFunction(cmd.GroupId))
+		}
```

Comment:
Well, it might be the same but... it's awkward to read and requires this comment and relies on the control-protocol library to continue behaving that way (not really documented).   I guess it's functionally ok but it gave me pause thinking... "what?  we ALWAYS fail?"

Reply:
I had it that way originally and was thinking of putting it back anyway.  If it made you pause as well I will go ahead and change it.